### ISSUE TITLE
Intended Dataflow for Mutations using GRDBQuery

### ISSUE THREAD

**Issue by DreamingInBinary** [groue/GRDBQuery#12]
In the demo project, the `Player` model is edited via transform in the `updateScore` function. For a more involved situation, say - we wanted to edit `Player`'s name, I'm trying to grasp the relationship between `Query` and mutations. I have it working in the demo app with this refactor to `PlayerFormView`:

```swift
struct PlayerFormView: View {
    @Environment(\.dbQueue) var dbQueue
    
    @State private var player: Player = Player.placeholder
    
    init(player:Player) {
        _player = State(initialValue: player)
    }
    
    var body: some View {
        VStack {
            TextEditor(text: $player.name)
            Stepper(
                "Score: \(player.score)",
                onIncrement: { player.score += 10 },
                onDecrement: { player.score = max(0, player.score - 10) })
            Spacer()
        }
        .onChange(of: player) { newValue in
            update()
        }
    }
    
    private func update() {
        do {
            _ = try dbQueue.write { db in
                try player.save(db)
            }
        } catch PersistenceError.recordNotFound {
            // Oops, player does not exist.
            // Ignore this error: `PlayerPresenceView` will dismiss.
            //
            // You can comment out this specific handling of
            // `PersistenceError.recordNotFound`, run the preview, change the
            // score, and see what happens.
        } catch {
            fatalError("\(error)")
        }
    }
}
```

The changes are basically swapping `var player: Player` for a `@State` version and calling `save` instead of `updateChanges`. This is driven by making `Player` conform to `Equatable` and then leveraging `onChange` to know when to save. Two questions:

1. Does this seem reasonable? Or is there are better suited way to architect this kind of mutation?
2. I assume that `Query` is _not_ meant for mutations, but rather observing changes. Is this correct? That is to say, we couldn't use the `@Query<PlayerPresenceRequest>` for a binding as well (I know this one is used to look for the status, but assume it simply vended a `Player` type).

**Update**
I guess another drawback with this is, if there is another window open - the `@Binding var player` doesn't seem to update in the windows which are not in the foreground (the nav bar does, but the `PlayerFormView` doesn't). This could also be an implementation error on my part:

![Simulator Screen Shot - iPad mini (6th generation) - 2022-02-18 at 19 03 55](https://user-images.githubusercontent.com/5566805/154779634-a0405a82-186f-4ed8-b93a-7b61f79c2816.png)

**Comment by groue** [comment]
Hello @DreamingInBinary,

An app that can display two distinct forms for editing the same data has to deal with multiple "sources of truth": database/file system updates, and user input from each visible form.

A similar setup can be found in text editors that support multiple editors for a single file. For example, in Xcode you can open the same file side by side. Keystrokes and editions performed in one pane automatically apply to the other pane. Select some text in one pane, edit that same text in the other pane, and see Xcode updates the selection. Edit and save the file from another application, and Xcode updates both editors.

Designing the dataflow in such a setup is certainly tricky. But the question is off-topic for this repository, I'm sorry.

There are simpler setups, such as designing an editing view that does not observe the database, and provides a "Done" button for explicit saving. This is the choice made by the [GRDBCombineDemo](https://github.com/groue/GRDB.swift/tree/master/Documentation/DemoApps/GRDBCombineDemo) demo app. Arguably, this demo app is iPhone-only (the UI is unable to display two editors for the same data).

**Comment by groue** [comment]
> ```swift
> do {
>     _ = try dbQueue.write { db in
>         try player.save(db)
>     }
> } catch PersistenceError.recordNotFound { }
> ```

I know this is not final code, but just in doubt: `save` can't throw PersistenceError.recordNotFound, since `save` inserts if the update fails (it [actually catches](https://github.com/groue/GRDB.swift/blob/v5.21.0/GRDB/Record/PersistableRecord.swift#L460) this error for this exact purpose). Generally speaking, only `updateXXX` methods can throw PersistenceError.recordNotFound.

**Comment by DreamingInBinary** [comment]
Thanks @groue - this is all super helpful! For that two way data flow, I ended up with something I'm good with. I think what helped me settle on a route was thinking of any `Query` as a read only data pipeline, I may get a model I throw in `@State` or something else from it, but I can't edit it.

For my latter thought, I also ended up using a class I made to wrap the Struct in so it updates in two windows, i.e.:

```
class WrappedStruct<T>: ObservableObject {
    @Published var item: T
    
    init(withItem item:T) {
        self.item = item
    }
}
```
